### PR TITLE
Package-Deps-Hash: add support for submodules

### PR DESCRIPTION
**Pull request by nickpape-msft**

### CODE REVIEW

**Comment by nickpape-msft** [comment]
Fixes issue #215

File: libraries/package-deps-hash/src/getPackageDeps.ts [hunk 1]
```diff
@@ -19,17 +19,24 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
   };
   const processOutputBlocks: string[] = stdout.split(PROCESS_OUTPUT_DELIMITER);
 
+  // A line is expected to look like:
+  // 100644 blob 3451bccdc831cb43d7a70ed8e628dcf9c7f888c8    src/typings/tsd.d.ts
+  // 160000 commit c5880bf5b0c6c1f2e2c43c95beeb8f0a808e8bac  web-build-tools
+  const gitRegex: RegExp = /([0-9]{6})\s(blob|commit)\s([a-f0-9]{40})\s*(.*)/;
+
```
Comment (qz2017) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: Need a new test case for submodules.

Comment (nickpape-msft) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: It's not really possible... The current test cases have a big problem where they are actually invoking the Git command on a local folder, which makes it difficult to represent a submodule.

I would be more in favor of doing a refactor on this code to make it be more modular and testable.

What do you think about that?

**Comment by nickpape-msft** [comment]
Ping @qz2017  @pgonzal

File: libraries/package-deps-hash/src/getPackageDeps.ts [hunk 2]
```diff
@@ -10,20 +10,26 @@ export function parseGitLsTree(output: string): Map<string, string> {
   const changes: Map<string, string> = new Map<string, string>();
 
   if (output) {
+    // A line is expected to look like:
+    // 100644 blob 3451bccdc831cb43d7a70ed8e628dcf9c7f888c8    src/typings/tsd.d.ts
+    // 160000 commit c5880bf5b0c6c1f2e2c43c95beeb8f0a808e8bac  web-build-tools
+    const gitRegex: RegExp = /([0-9]{6})\s(blob|commit)\s([a-f0-9]{40})\s*(.*)/;
```
Comment (qz2017) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: >6 [](start = 37, length = 1)

Is this always 6?

File: libraries/package-deps-hash/src/getPackageDeps.ts [hunk 3]
```diff
@@ -10,20 +10,26 @@ export function parseGitLsTree(output: string): Map<string, string> {
...
-      // A line is expected to look like:
-      // 100644 blob 3451bccdc831cb43d7a70ed8e628dcf9c7f888c8    src/typings/tsd.d.ts
-
       if (line) {
         // Take everything after the "100644 blob", which is just the hash and filename
-        const [hash, filename]: string[] = line.substr(line.indexOf('blob ') + 5).split('\t');
-        if (!filename || !hash) {
+        const matches: RegExpMatchArray = line.match(gitRegex);
+        if (matches) {
+          const hash: string = matches[3];
```
Comment (qz2017) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: >hash [](start = 16, length = 4)

Check if hash and filename have value before set.

File: libraries/package-deps-hash/src/getPackageDeps.ts (hunk 2 above)
Comment (nickpape-msft) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: Yes
The first 3 digits specify the type
Then the next 3 digits are unix file modes (r/w/x)
https://stackoverflow.com/questions/737673/how-to-read-the-mode-field-of-git-ls-trees-output/8347325#8347325

File: libraries/package-deps-hash/src/getPackageDeps.ts (hunk 3 above)
Comment (nickpape-msft) [inline libraries/package-deps-hash/src/getPackageDeps.ts]: OK

**Comment by qz2017** [comment]
:shipit:

**Pull request merged**